### DISCUSSION TITLE
[Support] Logger更新到V3后的问题。

### DISCUSSION THREAD

**Discussion by minefujiko9293** [midwayjs/midway#3482] [Q&A] [answered]
### Describe the problem（描述问题）

问题1：【Bug】
V3的新的配置结构，error类的Transport，配置文件名貌似不生效，感觉有bug，像我命名为 "WW_ES-error.log" ，但生成的文件还是旧的名字 "common-error.log" 。不管是在default还是clients中都不生效。file类的Transport好像是正常的。同时日志的文档，虽然说了v3的版本，好像跟 https://github.com/midwayjs/logger/blob/main/BREAKING-3.md 里面的说明也不太同步。

![image](https://github.com/midwayjs/midway/assets/23348616/c99bee0b-151a-462b-810b-f881499ede7b)

![image](https://github.com/midwayjs/midway/assets/23348616/e471af21-9f3d-4ca8-bd6f-af90b6439d3c)

同时，logger配置的代码类型提示好像也不太全，即使按照文档说的，在 src/interface.ts 中加入日志库的引用。
// src/interface.ts
import type {} from '@midwayjs/logger';
也不是很全

-------------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------------------------------

问题2：【优化】
Windows下，启动时控制台报错误。
[FileStreamRotator] Could not create symlink file:  E:\Workspace\saclglxt\midway-saclglxt-ww\tmps\logs/common-error.log  ->  common-error.log.2023-12-06

![image](https://github.com/midwayjs/midway/assets/23348616/905b26e1-e78b-4240-bf40-0feb04dd0557)

虽然v3的文档中已经说明了【windows 下不会生成软链】，但这个错误太刺眼啦。
![image](https://github.com/midwayjs/midway/assets/23348616/22732b2b-56c5-4393-ad4e-5e12c2f42b0f)

看看有没有能隐藏这个错误，毕竟每次保存代码都要刺眼下。

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed) 
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(3.13.5)
✓ @midwayjs/axios(not installed)      
✓ @midwayjs/bootstrap(3.13.5)
✓ @midwayjs/bull(3.13.5)
✓ @midwayjs/bull-board(3.13.5)        
✓ @midwayjs/cache(not installed)      
✓ @midwayjs/captcha(not installed)    
✓ @midwayjs/casbin(not installed)     
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)   
✓ @midwayjs/consul(not installed)     
✓ @midwayjs/core(3.13.5)
✓ @midwayjs/cos(not installed)        
✓ @midwayjs/cron(not installed)       
✓ @midwayjs/cross-domain(3.13.5)      
✓ @midwayjs/decorator(3.13.5)
✓ @midwayjs/etcd(not installed)       
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)       
✓ @midwayjs/grpc(not installed)       
✓ @midwayjs/http-proxy(not installed) 
✓ @midwayjs/i18n(3.13.5)
✓ @midwayjs/info(3.13.5)
✓ @midwayjs/jwt(3.13.5)
✓ @midwayjs/kafka(not installed)      
✓ @midwayjs/leoric(not installed)     
✓ @midwayjs/mikro(not installed)      
✓ @midwayjs/mock(3.13.5)
✓ @midwayjs/mongoose(not installed)   
✓ @midwayjs/oss(not installed)        
✓ @midwayjs/otel(not installed)       
✓ @midwayjs/passport(not installed)   
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed) 
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(3.13.5)
✓ @midwayjs/security(not installed)   
✓ @midwayjs/sequelize(not installed)  
✓ @midwayjs/session(3.13.5)
✓ @midwayjs/socketio(not installed)   
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)    
✓ @midwayjs/tablestore(not installed) 
✓ @midwayjs/tags(not installed)       
✓ @midwayjs/typegoose(not installed)  
✓ @midwayjs/typeorm(3.13.5)
✓ @midwayjs/upload(not installed)     
✓ @midwayjs/validate(3.13.5)
✓ @midwayjs/version(3.13.5)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)   
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)        
✓ @midwayjs/express(not installed)    
✓ @midwayjs/koa(3.13.5)
✓ @midwayjs/ws(not installed)
**********************************************************************      
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by minefujiko9293** [comment]
@czy88840616

**Reply by czy88840616** [reply]
windows 的处理的确是漏了

**Reply by minefujiko9293** [reply]
哥，问题1才是重点哈。error的日志名我改不了。怎么改都是‘common-error.log’。 新旧配置的写法写了也是不行

**Comment by minefujiko9293** [comment]
@czy88840616  刚刚再仔细测了下，
情况1： 
写法：整个midwayLogger配置对象旧写法
结果：【正确】自定义的WW_ES_Logger的error的文件名为 WW_ES-error.log
用例：
```
midwayLogger: {
      default: {
        dir: path.join(appInfo.appDir, 'tmps', 'logs'),
        maxFiles: '1d',
      },
      clients: {
        bullLogger: {
          consoleLevel: 'warn',
        },
        WW_ES_Logger: {
          dir: path.join(appInfo.appDir, 'tmps', 'logs', 'WW_ES'),
          fileLogName: 'WW_ES.log',
          errorLogName: 'WW_ES-error.log',
          maxFiles: '90d',
        },
      },
    },
```

情况2：
写法：遵循v3新写法
结果：【不符合预期】自定义的WW_ES_Logger的error的文件名不为WW_ES-error.log
用例：
```
 midwayLogger: {
      default: {
        transports: {
          console: {},
          file: {
            dir: path.join(appInfo.appDir, 'tmps', 'logs'),
            maxFiles: '1d',
          },
          error: {
            dir: path.join(appInfo.appDir, 'tmps', 'logs'),
            maxFiles: '7d',
          },
        },
      },
      clients: {
        bullLogger: {
          level: 'warn',
        },
        WW_ES_Logger: {
          transports: {
            file: {
              dir: path.join(appInfo.appDir, 'tmps', 'logs', 'WW_ES'),
              maxFiles: '90d',
              fileLogName: 'WW_ES.log',
            },
            error: {
              dir: path.join(appInfo.appDir, 'tmps', 'logs', 'WW_ES'),
              maxFiles: '90d',
              fileLogName: 'WW_ES-error.log',
            },
          },
        },
      },
    },
```

**Reply by czy88840616** [reply]
我测试下。

**Reply by czy88840616** [reply]
确认了，的确是bug。

**Reply by czy88840616** [reply]
https://github.com/midwayjs/logger/pull/109

**Reply by czy88840616** [reply, marked as answer]
已修复，试试 @midwayjs/logger@3.1.3

**Reply by minefujiko9293** [reply]
目前更新后简单测试了下暂时没有问题，后续有空时候我再测试下复制的情况。
剩下的问题就是Logger_V3版在config中的代码提示了，即使引入了
```
// src/interface.ts
import type {} from '@midwayjs/logger';
```
写具体的transports时，代码提示就只剩如下属性提示
![image](https://github.com/midwayjs/midway/assets/23348616/10400866-930c-4072-a675-6c7c8915fc7e)

好像缺这缺那，得看文档写配置了。
对刚入手v3的小伙伴可能懵了。

**Reply by czy88840616** [reply]
@midwayjs/logger@3.1.4，已恢复。

![image](https://github.com/midwayjs/midway/assets/418820/4f710170-1b59-4a3b-bcd6-ee4737ca2b2e)

**Reply by minefujiko9293** [reply]
试了ok，但是前提是得执行
```
// src/interface.ts
import type {} from '@midwayjs/logger';
```
当然这个说明只在 https://github.com/midwayjs/logger/blob/main/BREAKING-3.md  中提及。
建议高亮加入midway主文档中。或者什么时候稳定到不需要引入该代码。

刚刚看了原来v3的logger刚发布不久，幸苦了哥！
我也是为了升级midway核心版本，把我logger也升级才发现这么些问题的哈。

**Reply by czy88840616** [reply]
对的，这个是必须的，因为 ts 的 typings merge 必须先让编辑器 import 一次。

**Reply by minefujiko9293** [reply]
嗯嗯，我记忆中是之前v2版本时，不需要引入这段代码，也有代码提示（因为我的正式工程建的早，记忆中配置logger时有代码提示的）。
刚刚重新建了个新的项目，用的v2版本的logger，也需要引入才有提示了。
哈哈，记忆更乱了

**Reply by czy88840616** [reply]
因为要同时支持 v2 和 v3，框架代码不能内置 logger 库，3.13 变成可选依赖了，所以加载的时候全靠用户引用了。

框架的组件也是这么实现的，只是组件做了一个取巧的设计，在 configuration 中显式 import，即作为逻辑的一部分，又作为类型合并之用。
